### PR TITLE
Fix torch.remainder giving a zero result the dividend's sign

### LINKED ISSUE #193755: torch.remainder gives a zero result the sign of the dividend, not the divisor as documented
### 🐛 Describe the bug

`torch.remainder` is documented as giving a result with the same sign as the divisor:

> The result has the same sign as the divisor `other` and its absolute value is less than that of `other`.

When the result is zero it takes the sign of the dividend instead.

```python
import torch

r = torch.remainder(torch.tensor([-1.0]), torch.tensor([1.0]))
print(r.item(), r.signbit().item())  # -0.0 True

print(-1.0 % 1.0)  # 0.0, and the docs say this computes Python's modulus entrywise
```

The other direction too:

```python
r = torch.remainder(torch.tensor([2.0]), torch.tensor([-1.0]))
print(r.item(), r.signbit().item())  # 0.0 False, divisor is negative
```

NumPy and Python both give the zero the sign of the divisor. On the value set I test, 5 float32 pairs and 5 float64 pairs differ from NumPy this way, and every one of them is a signed zero.

The cause is that the sign fixup is guarded on the remainder being nonzero, in both the scalar and the vectorized path:

https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L396-L408

`std::fmod` returns a zero with the sign of the dividend, and nothing corrects it. NumPy's `npy_divmod` handles it with an explicit `mod = copysign(0, b)` in the zero case.

This is low severity on its own. It matters where the sign of a zero is then used, for example `copysign`, `signbit`, `atan2`, or dividing by the result.

Found by a differential tester. Related to #193753 and #193754, which are separate root causes in the same kernels.

### Versions

torch 2.13.0, CPU, macOS arm64.

cc @svekars @sekyondaMeta @AlannaBurke @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### PR DESCRIPTION
**Pull request by krrishapatel**
Fixes #193755

`torch.remainder` is documented to return a result with the sign of the divisor. A zero result came back with the sign of the dividend instead.

```python
>>> torch.remainder(torch.tensor([2.0]), torch.tensor([-1.0]))
tensor([0.])          # want -0., which is what Python and numpy give
>>> 2.0 % -1.0
-0.0
```

## Why

`fmod` gives a zero the sign of the dividend. The fixup that corrects the sign only ran when the result was nonzero, so zeros kept `fmod`'s sign. CPU and CUDA both do this. MPS computes `x - y * floor_divide(x, y)`, where the subtraction cancels to `+0.0` whatever the signs were.

## What changed

CPU scalar and vectorized loops, the CPU bfloat16 loop, CUDA, and MPS.

In the vectorized loops, keeping just the sign bit of `b` is `copysign(0, b)` and is cheaper than `Vectorized::copysign`, which calls into Sleef. That select also makes the nonzero term in the existing mask redundant, so the net cost is one extra op.

One build fix came with it. On machines without a full metal toolchain the shader is embedded into the binary by `write_metallib_headers.py`, but that step did not depend on the headers it embeds, so editing `c10/metal/utils.h` left a stale shader. `METAL_HEADER_DEPS` is now a dependency there, same as the `metal -> air` step already does.

## Testing

New `test_remainder_zero_sign` compares against Python's `%` for half, bfloat16, float and double, at lengths 1, 2, 3 and 64 so both the scalar and vectorized loops run. It checks `signbit` separately because `assertEqual` treats `-0.0` and `0.0` as equal.

Locally on arm64:
- 3360 dividend/divisor/length combinations per dtype match Python's `%` exactly, on CPU and MPS. Before: every case where the divisor was negative and the result was zero came out wrong.
- float32 and float64 match numpy bitwise on random data at lengths 1 through 1000.
- NaN, inf and zero divisors are unchanged. A zero divisor gives NaN, so it never reaches the new branch.
- `test_binary_ufuncs.py` 12715 passed, `test_ops.py -k "remainder or fmod"` 186 passed, `test_mps.py -k "remainder or fmod or floor_divide"` 38 passed.

CUDA is not compiled locally, so that line is untested. It mirrors the CPU change and `copysign(scalar_t(0), b)` resolves to the `float` overload for half and bfloat16, same as the existing call in `BinaryDivFloorKernel.cu`.

## Cost

Vectorized body, in-cache, interleaved in one binary: +0.8% float, +3.5% double. Whole op on 1M elements, single thread: +6% float32, +8% float64, +4% bfloat16, float16 unchanged. Using `Vectorized::copysign` instead was about twice that on double.

## Out of scope

- Inductor codegen keeps the `r != 0` guard and has the same bug. #175584 is already open on that file, so I left it alone to avoid conflicting.
- XPU has the same bug in `RemainderFloatingFunctor`, which lives in intel/torch-xpu-ops. The new test is `@skipXPU` with a pointer to it.

#193763 also touches `remainder_kernel`, for NaN handling. The two are independent, whichever lands second needs a trivial rebase.

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### CODE REVIEW

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!
